lib/parallel_test_base.pm

```diff
@@ -0,0 +1,1946 @@
+
+    assert_script_run("clear && ssh-keygen -b 2048 -t rsa -q -N \"\" -f ~/.ssh/id_rsa <<< y") if ($args{overwrite} == 1);
+    my $_ret = 0;
+    foreach my $_addr (split(/ /, $args{addr})) {
+        record_info("Config $_addr SSH PubKey auth");
+        next if (script_run("timeout --kill-after=3 --signal=9 15 ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_addr ls") == 0);
+        enter_cmd("clear", wait_still_screen => 3);
+        enter_cmd("ssh-copy-id -f -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -i ~/.ssh/id_rsa.pub root\@$_addr", wait_still_screen => 3);
+        if ($args{host} == 1) {
+            susedistribution::handle_password_prompt();
+        }
+        else {
+            check_screen("password-prompt", 60);
+            enter_cmd("novell", wait_screen_change => 50, max_interval => 1);
+            wait_still_screen(10);
+        }
+        my $_temp = 1;
+        $_temp = script_run("timeout --kill-after=3 --signal=9 15 ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_addr ls");
+        $_ret |= $_temp;
+        record_info("Guest $_addr SSH PubKeyAuth failed", "Can not establish ssh connection to guest $_addr using Public Key Authentication", result => 'fail') if ($_temp != 0);
+    }
+    croak("SSH public key authentication setup failed for certain system") if ($_ret != 0 and $args{die} == 1);
+    return $_ret;
+}
+
+=head2 check_host_os
+
+Check source and destination hosts have the same architecture, otherwise test run
+can not proceed.
+=cut
+
+sub check_host_architecture {
+    my ($self, %args) = @_;
+
+    record_info("Check host architecture");
+    my $_localip = get_var('LOCAL_IPADDR');
+    my $_localarch = script_output("uname -i");
+    my $_peerip = get_var('PEER_IPADDR');
+    my $_peerarch = script_output("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_peerip uname -i");
+    save_screenshot;
+    croak("Architecture $_localarch on $_localip does not match $_peerarch on $_peerip") if ($_localarch ne $_peerarch);
+}
+
+=head2 check_host_os
+
+Check source and destination hosts operating system version. Guest migration can
+not be done from the newer to the older. Main argument is host role (src or dst).
+=cut
+
+sub check_host_os {
+    my ($self, %args) = @_;
+    $args{role} //= 'src';
+
+    record_info("Check host os");
+    my $_ret = 0;
+    my $_localip = get_var('LOCAL_IPADDR');
+    my $_peerip = get_var('PEER_IPADDR');
+    if (is_sle) {
+        my ($_localosver, $_localossp,) = get_os_release;
+        my ($_peerosver, $_peerossp,) = get_os_release("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_peerip");
+        print "LOCAL OS $_localosver-sp$_localossp PEER OS $_peerosver-sp$_peerossp\n";
+        save_screenshot;
+        unless (($_peerosver > $_localosver) or ($_peerosver == $_localosver and $_peerossp >= $_localossp)) {
```

Comment:
This does not seem correct when $args{role} =dst ?

Reply:
This is correct. Please give you example if you think this is not correct.

Reply:
If role is dst, this function will directly return the initialized _ret 0 (means supported/correct) if peer/src is 12sp3 and local/dst is 15sp4.

Reply:
if role is dst, src host is 12sp3 and dst host is 15sp4, this should be returned 0, but here it will return 1 .

Reply:
I see, in later usage, it seems return value is not cared, the croak or not croak actually matters. That will be ok. But better differentiate return value between croak and not croak case.

Reply:
Return value is also OK. 0 is OK, 1 is NOTOK. The difference is whether croak.

Reply:
If role is src, src host is 12sp3 and dst host is 15sp4, the function is returning 0. So correct. 
If role is dst, src host is 12sp3 and dst host is 15sp4, the function is returning 1 and the record_info on Line 410 is giving correct info that it is valid, so conflicting info. "No croak" should match return 0, isn't it. You can also have a unit test in perl.

Besides, from just logic, if the design is to croak for invalid scenario, then it can happen when role is either src or dst. But now in this code, croak only happens to role=src. Problematic, isn't it?

Reply:
No. The recod_info is correct even with your example. For dst, there is no need to croak. Because dst does not have to migration back to src if src falls behind dst. But return value has to tell the real result. There is no problem here.

Reply:
Seems I was expecting differently. Doesn't matter. Up to your detailed design and need of course.